File: TypeScript/0155-min-stack.ts
// 2020-08-08 16:01:52

class MinStack {
  stack: number[] = []
  min: number[] = []

  constructor() {}

  push(x: number): void {
    this.stack.push(x)
    if (this.min.length == 0 || this.min[this.min.length - 1] >= x) {
      this.min.push(x)
    }
  }

  pop(): void {
    if (this.min[this.min.length - 1] == this.stack[this.stack.length - 1]) {
      this.min.pop()
    }
    this.stack.pop()
  }

  top(): number {
    return this.stack[this.stack.length - 1]
  }

  getMin(): number {
    return this.min[this.min.length - 1]
  }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: C++/0155_minstack.cc
// 2020-08-09 00:20:57

#include <vector>

using std::vector;

class MinStack {
public:
    /** initialize your data structure here. */
    MinStack() {

    }

    void push(int x) {
        if (min.empty() || min.back() >= x) {
            min.push_back(x);
        }
        stack.push_back(x);
    }

    void pop() {
        if (min.back() == stack.back()) {
            min.pop_back();
        }
        stack.pop_back();
    }

    int top() {
        return stack.back();
    }

    int getMin() {
        return min.back();
    }

private:
    vector<int> stack;
    vector<int> min;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */